fontforge/splinechar.c

```diff
@@ -620,25 +649,50 @@ return;
 	sp->nextcpdef = sp->nonextcp;
 	sp->prevcpdef = sp->noprevcp;
     } else if ( pointtype==pt_tangent ) {
-	if ( sp->next!=NULL && !sp->nonextcp && sp->next->knownlinear ) {
-	    sp->nextcp = sp->me;
-        SplineRefigure(sp->next);
-	} else if ( sp->prev!=NULL && !sp->nonextcp &&
-		BpColinear(&sp->prev->from->me,&sp->me,&sp->nextcp) ) {
-	    /* The current control point is reasonable */
-	} else {
-	    SplineCharTangentNextCP(sp);
-	    if ( sp->next ) SplineRefigure(sp->next);
-	}
-	if ( sp->prev!=NULL && !sp->noprevcp && sp->prev->knownlinear ) {
-	    sp->prevcp = sp->me;
-        SplineRefigure(sp->prev);
-	} else if ( sp->next!=NULL && !sp->noprevcp &&
-		BpColinear(&sp->next->to->me,&sp->me,&sp->prevcp) ) {
-	    /* The current control point is reasonable */
-	} else {
-	    SplineCharTangentPrevCP(sp);
-	    if ( sp->prev ) SplineRefigure(sp->prev);
+	if ( sp->next!=NULL && sp->prev!=NULL ) {
+		bigreal prevlinearity = Linearity(sp->prev);
+		bigreal nextlinearity = Linearity(sp->next);
+		if ( prevlinearity>=0 && nextlinearity>=0 ) {
+			if ( nextlinearity >= prevlinearity ) { /* make prev linear */
+				sp->prev->islinear = true;
+				sp->prev->from->nextcp = sp->prev->from->me;
+				sp->prevcp = sp->me; 
+				if ( sp->prev->from->pointtype!=pt_tangent) 
+					sp->prev->from->pointtype = pt_corner;
+				if ( sp->prev->order2 ) {
```

Comment:
Since review cycles are limited I'm dropping my "too much input to review" standard for this. 

My only question about the code at this stage is whether this part (with the `IntersectLines()`) is something already built into `SplineRefigureFixup()`, already used below in the curve case. Did you look at that?

Reply:
I don't think that I have considered `SplineRefigureFixup()` when I wrote this part. Is your question heading to the question whether the following block (and its "sp->next" counterpart)
```
				if ( sp->prev->order2 ) {
					BasePoint inter;
					if ( IntersectLines(&inter,&sp->me,&sp->prev->from->me,&sp->nextcp,&sp->next->to->me) ) {
						sp->nextcp = inter;
						sp->next->to->prevcp = inter;
						SplineRefigure(sp->prev);
					} /* else just leave things as they are */
				}
```
could be replaced by `SplineRefigureFixup(sp->prev);` (and `SplineRefigureFixup(sp->next);` respectively)? If yes, I will gladly check this.

Reply:
Yes -- and to be clear I don't know the answer, it just seemed like a possibility given a scan of that code and its use elsewhere in the function.

Reply:
I have checked it and it did not work:
![aa](https://user-images.githubusercontent.com/11213578/140412834-280837f3-2827-4383-912e-84dfa61e6db0.png)
left: original spline, middle: spline with top node converted to a tangent point with my original patch, right: using `SplineRefigureFixup(sp->prev);` and `SplineRefigureFixup(sp->next);`

Reply:
Alright - thanks for looking into this!